core/trino-main/src/main/java/io/trino/server/protocol/Query.java

```diff
@@ -791,26 +782,34 @@ private static StageStats toStageStats(StageInfo stageInfo)
                 .setPhysicalInputBytes(stageStats.getPhysicalInputDataSize().toBytes())
                 .setFailedTasks(stageStats.getFailedTasks())
                 .setCoordinatorOnly(stageInfo.isCoordinatorOnly())
-                .setSubStages(subStages.build())
-                .build();
-    }
+                .setNodes(countStageAndAddGlobalUniqueNodes(stageInfo, globalUniqueNodes));
 
-    private static Set<String> globalUniqueNodes(StageInfo stageInfo)
-    {
-        if (stageInfo == null) {
-            return ImmutableSet.of();
+        // Recurse into child stages to create their StageStats
+        List<StageInfo> subStages = stageInfo.getSubStages();
+        if (subStages.isEmpty()) {
+            builder.setSubStages(ImmutableList.of());
         }
-        ImmutableSet.Builder<String> nodes = ImmutableSet.builder();
-        for (TaskInfo task : stageInfo.getTasks()) {
-            // todo add nodeId to TaskInfo
-            URI uri = task.getTaskStatus().getSelf();
-            nodes.add(uri.getHost() + ":" + uri.getPort());
+        else {
+            ImmutableList.Builder<StageStats> subStagesBuilder = ImmutableList.builderWithExpectedSize(subStages.size());
+            for (StageInfo subStage : subStages) {
+                subStagesBuilder.add(toStageStats(subStage, globalUniqueNodes));
+            }
+            builder.setSubStages(subStagesBuilder.build());
         }
 
-        for (StageInfo subStage : stageInfo.getSubStages()) {
-            nodes.addAll(globalUniqueNodes(subStage));
+        return builder.build();
+    }
+
+    private static int countStageAndAddGlobalUniqueNodes(StageInfo stageInfo, Set<String> globalUniqueNodes)
+    {
+        List<TaskInfo> tasks = stageInfo.getTasks();
+        Set<String> stageUniqueNodes = new HashSet<>(tasks.size() + (tasks.size() / 2));
```

Comment:
Why not just `new HashSet<>(tasks.size())` ?

Reply:
HashSet/Map load factor is .75, but the initial capacity argument is the actual initial array size (not adjusted by the load factor). Passing the exact size of the number of unique elements being inserted would guarantee one rehash to occur.

Reply:
I see, I remembered that open hash sets from fastutils apply the load factor to the sizing hint and assumed that standard HashSet was doing the same.

Reply:
Yeah, that's one of those foot-guns to watch out for in the java collections library. Usually not a big deal, but if you're passing a size you're probably trying to avoid re-hashing and if you're not careful you're going to guarantee that rehashing occurs.